File: DynamicsAndControl/GoodVibrationsWithFreeball/TheFiniteElementMethod/Python/05_improved_bar_element.py
# https://www.youtube.com/watch?v=nGN9uqP67Ak&list=PL2ym2L69yzkaue8Ly2Oz51LALRzUV8LZ0&index=3

import numpy as np
import math

from scipy.linalg import eigh
from matplotlib import pyplot as plt
import time

def Bar2(num_elems):
  restrained_dofs = [0,]
  """  
  [m] = (p A l / 6) * 2   1
                      1   2


  [k] = E A /l * 1  -1
                -1   1
  """

  # element mass and stiffness matrices for a bar 
  # (2, 2)
  m = np.array([[2,  1], [1, 2]]) / (6. * float(num_elems))
  k = np.array([[1, -1], [-1, 1]]) * float(num_elems)

  # construct global mass and stiffness matrices
  # (num_elems + 1, num_elems + 1)
  M = np.zeros((num_elems + 1, num_elems + 1))
  K = np.zeros((num_elems + 1, num_elems + 1))

  # assembley of elements
  for i in range(num_elems):
    # print('num_elems: {} \ti: {}'.format(num_elems, i))
    #(num_elems + 1, num_elems + 1)
    M_temp = np.zeros((num_elems + 1, num_elems + 1))
    K_temp = np.zeros((num_elems + 1, num_elems + 1))

    M_temp[i : i + 2, i : i + 2] = m
    K_temp[i : i + 2, i : i + 2] = k

    M += M_temp
    K += K_temp

  # print("restrained_dofs")
  # print(restrained_dofs)
  # removed the fixed degree of freedom
  for dof in restrained_dofs:
    for i in [0, 1]:
      M = np.delete(M, dof, axis=i)
      K = np.delete(K, dof, axis=i)

  # eigen value problem
  evals, evecs = eigh(K, M)
  frequencies = np.sqrt(evals)

  return M, K, frequencies, evecs


def Bar3(num_elems):
  restrained_dofs = [0,]
  # element mass and stiffness matrices for a bar 
  # (3, 3)
  m = np.array([[4,  2, -1], [ 2, 16,  2], [-1,  2, 4]]) / (30.0 * float(num_elems))
  k = np.array([[7, -8,  1], [-8, 16, -8], [ 1, -8, 7]]) / 3.0 * float(num_elems)

  # construct global mass and stiffness matrices
  # (2 * num_elems + 1, 2 * num_elems + 1)
  M = np.zeros((2 * num_elems + 1, 2 * num_elems + 1))
  K = np.zeros((2 * num_elems + 1, 2 * num_elems + 1))

  # assembley of elements
  for i in range(num_elems):
    #(2 * num_elems + 1, 2 * num_elems + 1)
    M_temp = np.zeros((2 * num_elems + 1, 2 * num_elems + 1))
    K_temp = np.zeros((2 * num_elems + 1, 2 * num_elems + 1))

    first_node = 2 * i
    M_temp[first_node : first_node + 3, first_node : first_node + 3] = m
    K_temp[first_node : first_node + 3, first_node : first_node + 3] = k

    M += M_temp
    K += K_temp

  # print("restrained_dofs")
  # print(restrained_dofs)
  # removed the fixed degree of freedom
  for dof in restrained_dofs:
    for i in [0, 1]:
      M = np.delete(M, dof, axis=i)
      K = np.delete(K, dof, axis=i)

  # eigen value problem
  evals, evecs = eigh(K, M)
  frequencies = np.sqrt(evals)

  return M, K, frequencies, evecs


def PlotFrequencies(n):
  print("2 Node bar element")
  exact_frequency = math.pi / 2
  results = []

  for i in range(1, n):
    start = time.perf_counter()
    M, K, frequencies, evecs = Bar2(i)
    time_taken = time.perf_counter() - start
    error = (frequencies[0] - exact_frequency) / exact_frequency * 100.0
    results.append((i, error))

    print('Num Elems: {} \tFund. Frequency: {} \t Error: {}% \tShape: {} \tTime: {}'.format(i, round(frequencies[0], 3), round(error, 3), K.shape, round(time_taken*1000, 3)))


  print('Exact frequency: ', round(exact_frequency, 3))

  # plot the results
  elements = np.array([x[0] for x in results])
  errors2   = np.array([x[1] for x in results])

  print("3 Node bar element")
  exact_frequency = math.pi / 2
  results = []

  for i in range(1, n):
    start = time.perf_counter()
    M, K, frequencies, evecs = Bar3(i)
    time_taken = time.perf_counter() - start
    error = (frequencies[0] - exact_frequency) / exact_frequency * 100.0
    results.append((i, error))

    print('Num Elems: {} \tFund. Frequency: {} \t Error: {}% \tShape: {} \tTime: {}'.format(i, round(frequencies[0], 3), round(error, 3), K.shape, round(time_taken*1000, 3)))


  print('Exact frequency: ', round(exact_frequency, 3))

  # plot the results
  elements = np.array([x[0] for x in results])
  errors3   = np.array([x[1] for x in results])


  plt.plot(elements, errors2, 'o-')
  plt.plot(elements, errors3, 'o-')
  plt.xlim(elements[0], elements[-1])
  plt.xlabel('Number of Elements')
  plt.ylabel('Error (%)')
  plt.show()
if __name__ == '__main__':
  PlotFrequencies(11)

File: DynamicsAndControl/GoodVibrationsWithFreeball/TheFiniteElementMethod/CPP/src/05_improved_bar_element.cpp
#include "TheFiniteElementMethod/05_improved_bar_element.h"


ImprovedBarElement::ImprovedBarElement()
{
}

ImprovedBarElement::~ImprovedBarElement()
{

}

void ImprovedBarElement::ReduceDimention(const int& dof, MatrixXd& M)
{
  // M.block(dof, 0, M.rows() - 1, M.cols()) = M.bottomRows(M.rows() - 1);
  // M.block(0, dof, M.rows(), M.cols() - 1) = M.rightCols(M.cols() - 1);
  // M.conservativeResize(M.rows() - 1, M.cols() - 1);
  EigenUtils::RemoveRow(M, M.rows() - 1);
  EigenUtils::RemoveColumn(M, M.cols() - 1);
}

std::tuple<MatrixXd, MatrixXd, VectorXd, MatrixXd> ImprovedBarElement::Bar(const int num_elems)
{
  MatrixXd m, k;
  m.resize(3, 3);
  k.resize(3, 3);
  
  m <<  4,  2, -1,
        2, 16,  2,
       -1,  2,  4;

  k <<  7, -8,  1,
       -8, 16, -8,
        1, -8,  7;


  // element mass and stiffness matrices for a bar 
  m /= (30.0 * static_cast<double>(num_elems));
  k *= (1.0 / 3.0) * static_cast<double>(num_elems);

  // construct global mass and stiffness matrices
  MatrixXd M = MatrixXd::Zero(2 * num_elems + 1, 2 * num_elems + 1);
  MatrixXd K = MatrixXd::Zero(2 * num_elems + 1, 2 * num_elems + 1);

  // Assembly of elements
  for (int i = 0; i < num_elems; ++i)
  {
    int first_node = 2 * i;
    M.block(first_node, first_node, 3, 3) += m;
    K.block(first_node, first_node, 3, 3) += k;
  }

  // Restrained degrees of freedom
  std::vector<int> restrained_dofs = {0};
  for(int dof : restrained_dofs)
  {
    ReduceDimention(dof, M);
    ReduceDimention(dof, K);
  }

  // Eigenvalue problem: Solve the generalized eigenvalue problem
  Eigen::GeneralizedSelfAdjointEigenSolver<MatrixXd> es(K, M);
  VectorXd evals = es.eigenvalues();
  MatrixXd evecs = es.eigenvectors();

  // Calculate frequencies
  VectorXd frequencies = evals.cwiseSqrt();

  return std::make_tuple(M, K, frequencies, evecs);
}

void ImprovedBarElement::PlotFrequencies(const int n)
{
  double exact_frequency = M_PI / 2;
  // std::vector<std::pair<int, double>> results;
  std::vector<int> elements;
  std::vector<double> errors;

  
  for (int i = 1; i < n; ++i)
  {
    MatrixXd M, K;
    VectorXd frequencies;
    MatrixXd evecs;
    auto start = std::chrono::high_resolution_clock::now();
    std::tie(M, K, frequencies, evecs) = Bar(i);
    auto time_taken = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::high_resolution_clock::now() - start
        ).count();

    double error = (frequencies(0) - exact_frequency) / exact_frequency * 100.0;
    printf("Num Elems: %d, Fund. Frequency: %f, Error: %f, Time: %ld\n", 
      i, frequencies[0], error, time_taken);

    elements.emplace_back(i);
    errors.emplace_back(error);
  }

  plt::plot(elements, errors, "o-");
  plt::xlim(elements.at(0), elements.at(elements.size() - 1));
  plt::xlabel("Number of Elements");
  plt::ylabel("Error (%)");
  plt::show();
}

int main()
{
  ImprovedBarElement ibe;
  ibe.PlotFrequencies(11);
  
  return 0;
}